File: wwwroot/app/services/requestService.js
var RequestService = function() {

    var join = function(teamId, done, fail) {
        $.post("/request/join", { TeamId: teamId })
            .done(done)
            .fail(fail);
    };

    var invite = function (userId, done, fail) {
        $.post("/request/invite", { UserId: userId })
            .done(done)
            .fail(fail);
    };

    var accept = function(requestId, done, fail) {
        $.post("/request/accept", { RequestId: requestId })
            .done(done)
            .fail(fail);
    };

    var reject = function (requestId, done, fail) {
        $.post("/request/reject", { RequestId: requestId })
            .done(done)
            .fail(fail);
    };

    var cancel = function (requestId, done, fail) {
        $.post("/request/cancel", { RequestId: requestId })
            .done(done)
            .fail(fail);
    };

    var leave = function (done, fail) {
        $.post("/request/leave")
            .done(done)
            .fail(fail);
    };

    return {
        join: join,
        invite: invite,
        accept: accept,
        reject: reject,
        cancel: cancel,
        leave: leave
    };

}();

File: Core/Services/RequestService.cs
using System.Threading.Tasks;
using TeamUp.Core.Enums;
using TeamUp.Core.Models;

namespace TeamUp.Core.Services
{
    public class RequestService : IRequestService
    {
        private readonly IUnitOfWork _unitOfWork;

        public RequestService(IUnitOfWork unitOfWork)
        {
            _unitOfWork = unitOfWork;
        }

        public async Task Create(string userId, int teamId, RequestType requestType)
        {
            var newRequest = new Request(userId, teamId, requestType);
            await _unitOfWork.Requests.Add(newRequest);
            await _unitOfWork.CompleteAsync();
        }

        public async Task Accept(Request request)
        {

            if (request.Type == RequestType.FromMember)
            {
                var pendingRequests = await _unitOfWork.Requests.GetRequestsFromUser(request.UserId, RequestStatus.Pending);
                foreach (var req in pendingRequests)
                    req.SetStatus(RequestStatus.Canceled);
            }

            request.SetStatus(RequestStatus.Approved);

            var user = await _unitOfWork.Users.GetUserById(request.UserId, loadRelated: false);
            user.JoinTeam(request.TeamId);

            await _unitOfWork.CompleteAsync();
        }

        public async Task Reject(Request request)
        {
            request.SetStatus(RequestStatus.Rejected);
            await _unitOfWork.CompleteAsync();
        }

        public async Task Cancel(Request request)
        {
            request.SetStatus(RequestStatus.Canceled);
            await _unitOfWork.CompleteAsync();
        }
    }
}